packages/circuits/jwt-verifier-template.circom

```diff
@@ -0,0 +1,326 @@
+
+    // Verify if the typ in the header is "JWT"
+    var typLength = JWT_TYP_LENGTH();
+    var typ[typLength] = JWT_TYP();
+    signal typMatch[typLength] <== RevealSubstring(maxHeaderLength, typLength, 1)(header, jwtTypStartIndex, typLength);
+    for (var i = 0; i < typLength; i++) {
+        typMatch[i] === typ[i];
+    }
+
+    // Verify if the key `kid` in the header is unique
+    var kidKeyLength = JWT_KID_KEY_LENGTH();
+    var kidKey[kidKeyLength] = JWT_KID_KEY();
+    signal kidKeyMatch[kidKeyLength] <== RevealSubstring(maxHeaderLength, kidKeyLength, 1)(header, jwtKidStartIndex, kidKeyLength);
+    for (var i = 0; i < kidKeyLength; i++) {
+        kidKeyMatch[i] === kidKey[i];
+    }
+
+    // Reveal the kid in the header
+    signal kidStartIndex <== jwtKidStartIndex + kidKeyLength + 1;
+    var kidLength = JWT_KID_LENGTH();
+    signal kidBytes[kidLength] <== RevealSubstring(maxHeaderLength, kidLength, 0)(header, kidStartIndex, kidLength);
+    kid <== Hex2FieldModular(kidLength)(kidBytes);
+
+    // Verify if the key `iss` in the payload is unique
+    var issKeyLength = ISS_KEY_LENGTH();
+    var issKey[issKeyLength] = ISS_KEY();
+    signal issKeyMatch[issKeyLength] <== RevealSubstring(maxPayloadLength, issKeyLength, 1)(payload, issKeyStartIndex, issKeyLength);
+    for (var i = 0; i < issKeyLength; i++) {
+        issKeyMatch[i] === issKey[i];
+    }   
+
+    // Reveal the iss in the payload
+    signal issStartIndex <== issKeyStartIndex + issKeyLength + 1;
+    signal issMatch[issLen] <== RevealSubstring(maxPayloadLength, issLen, 0)(payload, issStartIndex, issLength);
+    iss <== Bytes2Ints(issLen)(issMatch);
+
+    // Verify if the key `iat` in the payload is unique
+    var iatKeyLength = IAT_KEY_LENGTH();
+    var iatKey[iatKeyLength] = IAT_KEY();
+    signal iatKeyMatch[iatKeyLength] <== RevealSubstring(maxPayloadLength, iatKeyLength, 1)(payload, iatKeyStartIndex, iatKeyLength);
+    for (var i = 0; i < iatKeyLength; i++) {
+        iatKeyMatch[i] === iatKey[i];
+    }
+
+    // Reveal the iat in the payload
+    var iatLength = TIMESTAMP_LENGTH();
+    signal iatStartIndex <== iatKeyStartIndex + iatKeyLength;
+    signal iatMatch[iatLength] <== RevealSubstring(maxPayloadLength, iatLength, 0)(payload, iatStartIndex, iatLength);
+    timestamp <== Digit2Int(iatLength)(iatMatch);
+
+    // Verify if the key `azp` in the payload is unique
+    var azpKeyLength = AZP_KEY_LENGTH();
+    var azpKey[azpKeyLength] = AZP_KEY();
+    signal azpKeyMatch[azpKeyLength] <== RevealSubstring(maxPayloadLength, azpKeyLength, 1)(payload, azpKeyStartIndex, azpKeyLength);
+    for (var i = 0; i < azpKeyLength; i++) {
+        azpKeyMatch[i] === azpKey[i];
+    }
+
+    // Reveal the azp in the payload
+    signal azpStartIndex <== azpKeyStartIndex + azpKeyLength + 1;
+    signal azpMatch[maxAzpLength] <== RevealSubstring(maxPayloadLength, maxAzpLength, 0)(payload, azpStartIndex, azpLength);
```

Comment:
Is it secure to skip the uniqueness check here?

Reply:
Same as the `kid` check.